op-service/sources/prefetching_eth_client.go

```diff
@@ -0,0 +1,249 @@
+	PayloadByHash(ctx context.Context, hash common.Hash) (*eth.ExecutionPayload, error)
+	PayloadByNumber(ctx context.Context, number uint64) (*eth.ExecutionPayload, error)
+	PayloadByLabel(ctx context.Context, label eth.BlockLabel) (*eth.ExecutionPayload, error)
+	FetchReceipts(ctx context.Context, blockHash common.Hash) (eth.BlockInfo, types.Receipts, error)
+	GetProof(ctx context.Context, address common.Address, storage []common.Hash, blockTag string) (*eth.AccountResult, error)
+	GetStorageAt(ctx context.Context, address common.Address, storageSlot common.Hash, blockTag string) (common.Hash, error)
+	ReadStorageAt(ctx context.Context, address common.Address, storageSlot common.Hash, blockHash common.Hash) (common.Hash, error)
+	Close()
+}
+
+type PrefetchingEthClient struct {
+	inner                 EthClientInterface
+	PrefetchingRange      uint64
+	PrefetchingTimeout    time.Duration
+	runningCtx            context.Context
+	runningCancel         context.CancelFunc
+	highestHeadRequesting uint64
+	highestHeadLock       sync.Mutex
+	wg                    *sync.WaitGroup // used for testing
+}
+
+// NewPrefetchingEthClient creates a new [PrefetchingEthClient] with the given underlying [EthClient]
+// and a prefetching range.
+func NewPrefetchingEthClient(inner EthClientInterface, prefetchingRange uint64, timeout time.Duration) (*PrefetchingEthClient, error) {
+	// Create a new context for the prefetching goroutines
+	runningCtx, runningCancel := context.WithCancel(context.Background())
+	return &PrefetchingEthClient{
+		inner:                 inner,
+		PrefetchingRange:      prefetchingRange,
+		PrefetchingTimeout:    timeout,
+		runningCtx:            runningCtx,
+		runningCancel:         runningCancel,
+		highestHeadRequesting: 0,
+	}, nil
+}
+
+func (p *PrefetchingEthClient) updateRequestingHead(start, end uint64) (newStart uint64, shouldFetch bool) {
+	// Acquire lock before reading/updating highestHeadRequesting
+	p.highestHeadLock.Lock()
+	defer p.highestHeadLock.Unlock()
+	if start <= p.highestHeadRequesting {
+		start = p.highestHeadRequesting + 1
+	}
+	if p.highestHeadRequesting < end {
+		p.highestHeadRequesting = end
+	}
+	return start, start <= end
+}
+
+func (p *PrefetchingEthClient) FetchWindow(start, end uint64) {
+	if p.wg != nil {
+		p.wg.Add(1)
+		defer p.wg.Done()
+	}
+
+	start, shouldFetch := p.updateRequestingHead(start, end)
+	if !shouldFetch {
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(p.runningCtx, p.PrefetchingTimeout)
+	defer cancel()
+	for i := start; i <= end; i++ {
+		p.FetchBlockAndReceipts(ctx, i)
+	}
+}
+
+func (p *PrefetchingEthClient) FetchBlockAndReceipts(ctx context.Context, number uint64) {
+	// Return data and error info is discarded as we are just filling the inner cache
+	blockInfo, _, err := p.inner.InfoAndTxsByNumber(ctx, number) // We need err here, though, to make sure blockInfo is safe to access
+	if err != nil {
+		// It's unsafe to access blockInfo. Return.
```

Comment:
This commenting here is a bit verbose, I suggest removing
```suggestion
	blockInfo, _, err := p.inner.InfoAndTxsByNumber(ctx, number)
	if err != nil {
```